### ISSUE TITLE
Return value from const_trait_impl did not promote to constant?

### ISSUE THREAD

**Issue by crlf0710** [rust-lang/rust#69574]
I tried this code:

```rust
#![feature(const_trait_impl)]
#![feature(const_fn)]
use std::marker::PhantomData;

trait ConstDefault {
    fn const_default() -> Self;
}

impl const ConstDefault for u8 {
    fn const_default() -> Self { 0 }
}

const fn foo() -> &'static u8 {
    &u8::const_default()
}
```

I expected to see this happen: This should pass compile if i'm not mistaken.

Instead, this happened: The const value promotion did not happen.

cc @ecstatic-morse @oli-obk

**Comment by ecstatic-morse** [comment]
The result of `const fn` calls do not get promoted. See rust-lang/const-eval#19 for background.

**Comment by crlf0710** [comment]
@ecstatic-morse 
Thank you for the pointer! Yes after reading that, it seems reasonable to prohibit this promotion here.
I believe for the above case, it would be nice to have a separate error code and explain this to the user. Also instruction user to add a constant item there to workaround this limitation, e.g.
```rust
 const V: u8 = u8::const_default();
 &V
```

@estebank

**Comment by ecstatic-morse** [comment]
@crlf0710 I totally agree regarding the diagnostics. I think it will be a bit tricky to implement however. Perhaps it is better to have a dedicated issue for this? I only found #66541 which seems to have arrived at a similar conclusion.

**Comment by crlf0710** [comment]
Sure, feel free to edit the text of this issue or open another issue etc, just whatever is helpful to the teams and work groups. (i don't really know what works the best.) It's not urgent at all since the workaround is so easy after the rule is learnt (just explicitly write an constant item, instead of expecting automatic promotion).

**Comment by oli-obk** [comment]
We'd have to register code that can be promoted via an explicit constant somewhere where borrowck can access it again. All the ways that I can think of would make this either resource consuming in the happy path or very convoluted in the error reporting path.

While I do agree it would be nice to have a better diagnostic, that seems to be little gain for a lot of effort for now.

But going back to the original issue. This should indeed compile I think because you are promoting from inside a const fn where all the const rules need to apply anyway. I though we were promoting anything const inside const fn. Or are we just doing that inside constants?

**Comment by ecstatic-morse** [comment]
@oli-obk Just inside constants, since they always execute at compile-time.

**Comment by oli-obk** [comment]
Ok, so basically, when a lifetime error occurs, we should rerun the promotion check with const rules enabled, and if they succeed, suggest to insert an intermediate constant. Additionally, due to the fact that we don't have `const {}` blocks or generic constants, explain this limitation in the presence of generics in the to-be-promoted code.

**Comment by estebank** [comment]
The above comment is still needed and likely the best course of action.

[Current output](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2024&gist=c4ac1441164a040d4a9112d08c449125):

```
error[E0515]: cannot return reference to temporary value
  --> src/lib.rs:13:5
   |
13 |     &u8::const_default()
   |     ^-------------------
   |     ||
   |     |temporary value created here
   |     returns a reference to data owned by the current function
```
